sync-core/src/main/java/com/cloudant/http/HttpConnection.java

```diff
@@ -133,56 +155,82 @@ public void setRequestBody(InputStream input, long inputLength) {
-                connection.setFixedLengthStreamingMode((int)this.inputLength);
-            } else {
-                // TODO some situations where we can't do chunking, like multipart/related
-                /// https://issues.apache.org/jira/browse/COUCHDB-1403
-                connection.setChunkedStreamingMode(1024);
-            }
 
-            // See "8.2.3 Use of the 100 (Continue) Status" in http://tools.ietf.org/html/rfc2616
-            // Attempting to write to the connection's OutputStream may cause an exception to be
-            // thrown. This is useful because it avoids sending large request bodies (such as
-            // attachments) if the server is going to reject our request. Reasons for rejecting
-            // requests could be 401 Unauthorized (eg cookie needs to be refreshed), etc.
-            connection.setRequestProperty("Expect", "100-continue");
+        boolean retry = true;
+        int n = numberOfRetries;
+        while (retry && n-- > 0) {
 
-            int bufSize = 1024;
-            int nRead = 0;
-            byte[] buf = new byte[bufSize];
-            InputStream is = input;
-            OutputStream os = connection.getOutputStream();
+            //set the response code to -1 again, to cover retries.
+            this.responseCode = -1;
 
-            while((nRead = is.read(buf)) >= 0) {
-                os.write(buf, 0, nRead);
+            System.setProperty("http.keepAlive", "false");        
+            connection = (HttpURLConnection) url.openConnection();
+            for (String key : requestProperties.keySet()) {
+                connection.setRequestProperty(key, requestProperties.get(key));
             }
-            os.flush();
-            // we do not call os.close() - on some JVMs this incurs a delay of several seconds
-            // see http://stackoverflow.com/questions/19860436
+            // always read the result, so we can retrieve the HTTP response code
+            connection.setDoInput(true);
+            connection.setRequestMethod(requestMethod);
+            if (contentType != null) {
+                connection.setRequestProperty("Content-type", contentType);
+            }
+            if (url.getUserInfo() != null) {
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                OutputStream bos = Base64OutputStreamFactory.get(baos);
+                bos.write(url.getUserInfo().getBytes());
+                bos.flush();
+                bos.close();
+                String encodedAuth = baos.toString();
+                connection.setRequestProperty("Authorization", String.format("Basic %s", encodedAuth));
+            }
+            
+            HttpConnectionFilterContext currentContext = new HttpConnectionFilterContext(this);
+            for (HttpConnectionRequestFilter requestFilter : requestFilters) {
+                currentContext = requestFilter.filterRequest(currentContext);
+            }        
```

Comment:
Response filters probably shouldn't be doing anything that would cause exceptions, so I think its fine to bomb out if a runtime exception gets thrown. The author of any implementation should handle exceptions that arise.